File: files/CH-041.py
import pandas as pd

input = pd.read_excel("CH-041 Transformation.xlsx", usecols="B:E", skiprows=1, nrows = 9)
test = pd.read_excel("CH-041 Transformation.xlsx", usecols="G:H", skiprows=1, nrows = 20)

result = input.copy()
# name columns
result.columns = ["Machinary Code", "Col1", "Col2", "Col3"]
result = result.melt(id_vars=["Machinary Code"], var_name="col", value_name="Product Code").dropna(subset=["Product Code"])
result = result.sort_values(by=["Product Code", "col"])
result = result.drop(columns=["col"]).reset_index(drop=True)

print(result.equals(test)) # True

File: files/CH-041.R
library(tidyverse)
library(readxl)

input = read_excel("files/CH-041 Transformation.xlsx", range = "B2:E11")
test  = read_excel("files/CH-041 Transformation.xlsx", range = "G2:H21")

result = input %>%
  select(`Machinary Code` = 1, Col1 = 2, Col2 = 3, Col3 = 4) %>%
  pivot_longer(cols = -c(1), names_to = "col", values_to = "Product Code", values_drop_na = TRUE) %>%
  arrange(str_extract(`Product Code`, "\\d+"), col) %>%
  select(-col)

identical(result, test)
# [1] TRUE
